pkg/plugins/manager/installer/installer.go

```diff
@@ -0,0 +1,628 @@
+package installer
+
+import (
+	"archive/zip"
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/plugins/manager/installer/logger"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
+
+type Installer struct {
+	retryCount int
+
+	httpClient          http.Client
+	httpClientNoTimeout http.Client
+	grafanaVersion      string
+	log                 logger.Logger
+}
+
+const (
+	permissionsDeniedMessage = "could not create %q, permission denied, make sure you have write access to plugin dir"
+)
+
+var (
+	ErrNotFoundError = errors.New("404 not found error")
+	reGitBuild       = regexp.MustCompile("^[a-zA-Z0-9_.-]*/")
+)
+
+type BadRequestError struct {
+	Message string
+	Status  string
+}
+
+func (e *BadRequestError) Error() string {
+	if len(e.Message) > 0 {
+		return fmt.Sprintf("%s: %s", e.Status, e.Message)
+	}
+	return e.Status
+}
+
+func New(skipTLSVerify bool, grafanaVersion string, logger logger.Logger) *Installer {
+	return &Installer{
+		httpClient:          makeHttpClient(skipTLSVerify, 10*time.Second),
+		httpClientNoTimeout: makeHttpClient(skipTLSVerify, 10*time.Second),
+		log:                 logger,
+		grafanaVersion:      grafanaVersion,
+	}
+}
+
+// Install downloads the plugin code as a zip file from specified URL
+// and then extracts the zip into the provided plugins directory.
+func (i *Installer) Install(pluginID, version, pluginsDir, pluginZipURL, pluginRepoURL string) error {
+	isInternal := false
+
+	var checksum string
+	if pluginZipURL == "" {
+		if strings.HasPrefix(pluginID, "grafana-") {
+			// At this point the plugin download is going through grafana.com API and thus the name is validated.
+			// Checking for grafana prefix is how it is done there so no 3rd party plugin should have that prefix.
+			// You can supply custom plugin name and then set custom download url to 3rd party plugin but then that
+			// is up to the user to know what she is doing.
+			isInternal = true
+		}
+		plugin, err := i.getPluginMetadataFromPluginRepo(pluginID, pluginRepoURL)
+		if err != nil {
+			return err
+		}
+
+		v, err := selectVersion(&plugin, version)
+		if err != nil {
+			return err
+		}
+
+		if version == "" {
+			version = v.Version
+		}
+		pluginZipURL = fmt.Sprintf("%s/%s/versions/%s/download",
+			pluginRepoURL,
+			pluginID,
+			version,
+		)
+
+		// Plugins which are downloaded just as sourcecode zipball from github do not have checksum
+		if v.Arch != nil {
+			archMeta, exists := v.Arch[osAndArchString()]
+			if !exists {
+				archMeta = v.Arch["any"]
+			}
+			checksum = archMeta.SHA256
+		}
+	}
+
+	i.log.Debug(fmt.Sprintf("Installing plugin\nfrom: %s\ninto: %s\n\n", pluginZipURL, pluginsDir))
```

Comment:
Nit. Might want to add a Debugf method to remove the fmt.Sprintf? Applies to similar Info, Warn, Error, Success and Failure methods as well.